### ISSUE TITLE
Generics

### ISSUE THREAD

**Issue by tomjoju** [nus-cs2030/2021-s1#530]
# Description 
Hi all, just wanted to understand the concepts of covariance and contravariance better. 
```
List<? extends Number> myCov = new ArrayList<Double>();
List<? super Integer> myCont = new ArrayList<Integer>();
Number n = myCov.get(0); //works fine
Integer myNum = myCont.get(0); //compiler-error
myCov.add(45); //compiler error
myCont.add(45); //works fine
```
I understand that line 3 works fine because we are sure that whatever we extract from `myCov` can be cast into a `Number`. Similarly, contravariant structures will not allow you to retrieve items as you cannot be sure that an item you obtain from `myCont` can be cast to an `Integer`. But, I could not fully understand the explanation of why we cannot put items into a covariant structure (line 5) and why we are allowed to put items into a contravariant structure (line 6). Could someone explain it to me in a simple way? Thank you :)

## Topic: 
Contravariance and Covariance

**Comment by benluiwj** [comment]
sorry does line 2 compile? isn't Double a child-class of Number? I tried to do it in jshell and it straight away failed...

**Comment by benluiwj** [comment]
Line 2 doesn't compile right? cos Double is not a super class of Number

**Comment by tomjoju** [comment]
@benluiwj Yes, you are right. I have edited that now. Thanks for pointing that out :)

**Comment by benluiwj** [comment]
I think its cos of the auto box and unboxing thing. So 45 is an int which would be autoboxed into Integer. I think it would be a different type from double so there will be an error. I may be wrong

**Comment by benluiwj** [comment]
Actually it doesn't only work all the time for contravariance as well. `List<? super Integer> lst = new ArrayList<Object>()` will not allow you to add Objects into the list as well (same error as your covariant one here as well)

**Comment by benluiwj** [comment]
https://llorllale.github.io/java-variance/#arrays-in-java-have-always-been-covariant

This explains it to a certain extent

**Comment by tlylt** [comment]
gonna expand on this...

So there are two cases.
1. upper bound wildcard parameterized type ( `<? extends Number>`)
2. lower bound wildcard parameterized type    ( `<? super Integer>`)

What is accessible depends on the information retained in the type.
---

#### Case 1 `List<? extends Number> myCov = new ArrayList<Double>();`
- take

myCov knows it has an unknown type that has an upper bound of `Number`.
Then because of this fact of upper bounded wildcard:
>**A returned object of "unknown" type is known to be compatible to the upper bound.**

`Number n = myCov.get(0);` works (p.s. provided myCov is not empty)
- put

However, because we cannot tell what is the unknown that is inside myCov, could be `Integer`, could be `Double`, so we cannot put anything in, except a null.

#### Case 2 `List<? super Integer> myCont = new ArrayList<Integer>();`
- take

myCont knows it has an unknown type that has a lower bound of `Integer`
This information is not enough to determine if `Integer` is taken out of List, hence you are generally not able to do a `.get`, except for, if you now realized,  `Object o = myCont.get(0)`, (p.s. provided not empty)

- put

The lower bound does provide you with the information that if you put an `Integer` into the List, there is nothing wrong.
But we cannot pass a `Number` as an argument, because the compiler does not know whether the `List<? super Integer>` refers to a `List<Number>` or perhaps to a `List<Comparable<Integer>>` , in which case a Number will be unacceptable because it is not comparable. 

References:
Generics [FAQ here](http://www.angelikalanger.com/GenericsFAQ/FAQSections/TechnicalDetails.html#Which%20methods%20that%20use%20the%20type%20parameter%20in%20the%20argument%20or%20return%20type%20are%20accessible%20in%20an%20upper%20bound%20wildcard%20instantiation?) and [FAQ here](http://www.angelikalanger.com/GenericsFAQ/FAQSections/TechnicalDetails.html#Which%20methods%20that%20use%20the%20type%20parameter%20in%20the%20argument%20or%20return%20type%20are%20accessible%20in%20a%20lower%20bound%20wildcard%20instantiation?)

**Comment by tomjoju** [comment]
Thanks so much guys @benluiwj @tlylt . This concept is a lot clearer to me now :) Just to confirm, @tlylt so if I declare a `List<? super Integer>`, I can definitely put in `Integer` values but any other super classes are not allowed because the compiler does not know the actual data type of the elements in the list?

**Comment by tlylt** [comment]
Apologies, I made a typo there. Let me correct it and try to explain this a bit more. Have to admit that the more you think about it, the more confusing it gets haha...

---
#### Case 2 `List<? super Integer> myCont = new ArrayList<Integer>();`
- put

>"The lower bound does provide you with the information that if you put an `Integer` into the List, there is nothing wrong."

=> Because `Integer` is the lower bound, replacing `?` with `Integer`, you get `List<Integer super *whatever super of Integer*>` which will always be valid. In fact, `List<? super Number>` means you can put `Number` and any subtype of `Number`, 
like `Integer` and `Long` into it!!!

Strange right, one wonders "how come?? they are not super of Number". 
But if we think of it this way, it makes sense:
" `Integer` and `Long` are always going to be the child of any super type of `Number`,  
e.g. `Integer` and `Long` are children of `Object` which is the supertype of `Number`"

Big brain moment 
![image](https://user-images.githubusercontent.com/41845017/100529225-266dbb80-3220-11eb-986c-ed8d0986a760.png)


>But we cannot pass a `Number` as an argument, because the compiler does not know whether the `List<? super Integer>` refers to a  **`List<Integer>`**  or perhaps to a `List<Comparable<Integer>>` , in which case a Number will be unacceptable because it is not comparable. 

It should be that the compiler does not know whether the `List<? super Integer>` refers to a  **`List<Number>`**  or perhaps to a `List<Comparable<Integer>>` . 
- `Number` is a class that fits the requirement. i.e. `List<Number super Integer>`
- `Comparable<Integer>` is an interface that fits the requirement, i.e. `List<Comparable<Integer> super Integer>`
This means the following works:
```
List<? super Integer> n = new ArrayList<Number>();
List<? super Integer> n = new ArrayList<Comparable<Integer>>();
```
So in the two possible cases above, what is the common type that can go into the List?
The answer is only `Integer`, because
 `Number` should not go into `ArrayList<Comparable<Integer>>`
and 
 `Comparable<Integer>` should not go into `ArrayList<Number>`

But the compiler won't know which is which, because all the compiler knows is this is `List<? super Integer>`

---
One place where confusion might sneak in is: putting stuff in is different from assigning the List,
So we can only have supertype of Integer on the Right Hand Side of assignment:
```
List<? super Integer> lst = new ArrayList<Number>();
List<? super Integer> lst = new ArrayList<Comparable<Integer>>();
```
But we can put any subtype of Integer into the assigned List
```
class A extends Integer{}
lst.put(1)
lst.put(new A()) // also works
```

**Comment by pikasean** [comment]
Hey guys! I was doing some research on generics too and I came across this illustration which I found to be quite helpful in enhancing my understanding of the subject. Picture courtesy of [Andrey Tyukin](https://stackoverflow.com/users/2707792/andrey-tyukin).

![image](https://user-images.githubusercontent.com/53890217/100542068-e7248680-3282-11eb-8add-32bb57fc8200.png)
